File: admin/test/batch-code.test.mjs
import assert from "node:assert/strict";
import test from "node:test";
import {
  batchCodeDraftForDialog,
  defaultBatchCode,
} from "../src/lib/batch-code.ts";

const fixedDate = new Date("2026-02-03T04:05:06.007Z");

test("uses the house name and farm-local minute in new batch drafts", () => {
  assert.equal(defaultBatchCode("东一舍", fixedDate), "东一舍-20260203-1205");
});

test("keeps a manual batch code while the dialog remains open", () => {
  assert.equal(
    batchCodeDraftForDialog("人工批次-复配", false, "东一舍", fixedDate),
    "人工批次-复配",
  );
});

test("normalizes separators and falls back for a blank house name", () => {
  assert.equal(
    defaultBatchCode("  东一 / 舍--A  ", fixedDate),
    "东一-舍-A-20260203-1205",
  );
  assert.equal(defaultBatchCode(" /_- ", fixedDate), "兔舍-20260203-1205");
});

test("truncates long names to the backend batch-code limit", () => {
  const code = defaultBatchCode("超长兔舍".repeat(30), fixedDate);
  assert.equal(Array.from(code).length, 100);
  assert.equal(code.endsWith("-20260203-1205"), true);
});

File: app/test/domain/batches/batch_code_test.dart
import 'package:flutter_test/flutter_test.dart';
import 'package:rabbit_flutter/src/domain/batches/batch_code.dart';

void main() {
  final fixedTime = DateTime.utc(2026, 2, 3, 4, 5, 6, 7);

  test('uses the house name and farm-local minute in a new batch code', () {
    expect(defaultBatchCode('东一舍', fixedTime), '东一舍-20260203-1205');
  });

  test('the same instant produces the same farm timestamp', () {
    expect(
      defaultBatchCode('东一舍', fixedTime),
      defaultBatchCode('东一舍', fixedTime.toLocal()),
    );
  });

  test('normalizes separators and falls back for a blank house name', () {
    expect(
      defaultBatchCode('  东一 / 舍--A  ', fixedTime),
      '东一-舍-A-20260203-1205',
    );
    expect(defaultBatchCode(' /_- ', fixedTime), '兔舍-20260203-1205');
  });

  test('truncates long names to the backend batch-code limit', () {
    final code = defaultBatchCode('超长兔舍' * 30, fixedTime);

    expect(code.runes.length, maxBatchCodeLength);
    expect(code, endsWith('-20260203-1205'));
  });
}
